pilot/pkg/model/context.go

```diff
@@ -849,14 +858,35 @@ func (node *Proxy) SetServiceInstances(serviceDiscovery ServiceDiscovery) {
 	node.ServiceInstances = instances
 }
 
-// SetWorkloadLabels will set the node.Metadata.Labels only when it is nil.
+// SetWorkloadLabels will set the node.Labels.
+// It merges both node meta labels and workload labels and give preference to workload labels.
 func (node *Proxy) SetWorkloadLabels(env *Environment) {
-	// First get the workload labels from node meta
-	if len(node.Metadata.Labels) > 0 {
+	if len(node.Labels) > 0 {
 		return
 	}
-	// Fallback to calling GetProxyWorkloadLabels
-	node.Metadata.Labels = env.GetProxyWorkloadLabels(node)
+	labels := env.GetProxyWorkloadLabels(node)
+	// when IstioMetaLabels unset, calculate the IstioMetaLabels. It is only set once.
+	if node.IstioMetaLabels == nil {
+		IstioMetaLabels := make(map[string]string, len(node.Metadata.Labels))
+		for k, v := range node.Metadata.Labels {
+			IstioMetaLabels[k] = v
+		}
+		for k := range labels {
+			delete(IstioMetaLabels, k)
+		}
+		node.IstioMetaLabels = IstioMetaLabels
+	}
+
+	node.Labels = make(map[string]string, len(labels)+len(node.IstioMetaLabels))
+	// we can't just equate proxy workload labels to node meta labels as it may be customized by user
+	// with `ISTIO_METAJSON_LABELS` env (pkg/bootstrap/config.go extractAttributesMetadata).
+	// so, we fill the `ISTIO_METAJSON_LABELS` as well.
+	for k, v := range node.IstioMetaLabels {
+		node.Labels[k] = v
+	}
+	for k, v := range labels {
```

Comment:
are we missing label removal here? Seems node metadata labels cannot be deleted?

Reply:
`IstioMetaLabels` exist here to help processing pod label delete, because it caches the `ISTIO_METAJSON_LABELS`